components/object_store/src/multipart.rs

```diff
@@ -0,0 +1,258 @@
+            self.completed_parts[part_idx] = Some(part);
+        }
+        Ok(())
+    }
+}
+
+/// [ Note: Methods in this impl are added by ceresdb, not included in the
+/// `object_store` crate ].
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    /// Send all buffer data to object store in final stage.
+    fn final_flush_buffer(mut self: Pin<&mut Self>) {
+        let part_size = self.part_size;
+        while !self.current_buffer.is_empty() {
+            let size = part_size.min(self.current_buffer.len());
+            let out_buffer = self.current_buffer.drain(0..size).collect::<Vec<_>>();
+
+            let inner = Arc::clone(&self.inner);
+            let part_idx = self.current_part_idx;
+            self.tasks.push(Box::pin(async move {
+                let upload_part = inner.put_multipart_part(out_buffer, part_idx).await?;
+                Ok((part_idx, upload_part))
+            }));
+            self.current_part_idx += 1;
+        }
+    }
+
+    /// Send buffer data to object store in write stage.
+    fn flush_buffer(mut self: Pin<&mut Self>) {
+        let part_size = self.part_size;
+
+        // We will continuously submit tasks until size of the buffer is smaller than
+        // `part_size`.
+        while self.current_buffer.len() >= part_size {
+            let out_buffer = self.current_buffer.drain(0..part_size).collect::<Vec<_>>();
+            let inner = Arc::clone(&self.inner);
+            let part_idx = self.current_part_idx;
+            self.tasks.push(Box::pin(async move {
+                let upload_part = inner.put_multipart_part(out_buffer, part_idx).await?;
+                Ok((part_idx, upload_part))
+            }));
+            self.current_part_idx += 1;
+            // *enough_to_send = self.current_buffer.len() >= part_size;
+        }
+    }
+}
+
+/// The process of ObjectStore write multipart upload is:
+/// First, Call ObjectStore::multi_upload to begin multipart upload.
+/// Second, Call AsyncWrite::poll_write to wrtie buffer data to oject,this
+/// fuction can be called many times. Last, Call AsyncWrite::poll_shutdown to
+/// finish current mulipart upload. See more in
+/// [`analytic_engine::sst::parquet::writer::ParquetSstWriter::write`]
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    // The `poll_flush` function will only flush the in-progress tasks.
+    // The `final_flush` method called during `poll_shutdown` will flush
+    // the `current_buffer` along with in-progress tasks.
+    fn final_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
+
+        // If current_buffer is not empty, see if it can be submitted
+        if !self.current_buffer.is_empty() && self.tasks.len() < self.max_concurrency {
```

Comment:
Let the `final_flush_buffer` to do `current_buffer` empty check:
```suggestion
        if self.tasks.len() < self.max_concurrency {
```